### PR TITLE
Prevent the character to go out of bound in animation viewport

### LINKED ISSUE #7646: Animation Editor: Grid size if fixed, it should be infinite
When characters are running they may run out of the grid. Move the grid with the character

### PR DESCRIPTION
**Pull request by amzn-rhhong**
Resolves #7646

Signed-off-by: rhhong <rhhong@amazon.com>

### CODE REVIEW

File: Gems/AtomLyIntegration/EMotionFXAtom/Code/Tools/EMStudio/AnimViewportRenderer.cpp [hunk 1]
```diff
@@ -224,6 +224,22 @@ namespace EMStudio
...
+        // Used the projected position to check the distance.
+        AZ::Vector3 characterProjectedPos = GetCharacterCenter();
+        characterProjectedPos.SetZ(groundPos.GetZ());
+
+        float distance = groundPos.GetDistanceEstimate(characterProjectedPos);
+        if (distance > BoundMaxDistance)
+        {
+            AZ::TransformBus::Event(m_groundEntity->GetId(), &AZ::TransformBus::Events::SetWorldTranslation, characterProjectedPos);
+        }
+    }
```
Comment (amzn-jillich) [inline Gems/AtomLyIntegration/EMotionFXAtom/Code/Tools/EMStudio/AnimViewportRenderer.cpp:241]: Doesn't this result in pops when it actually moves the ground?

What I would do is something like this:

```
const float tileSize = X; // We need to find out this value.
const float newX = std::fmod(charPos.x, tileSize);
const float newY = std::fmod(charPos.y, tileSize);
groundPlane.SetPos(newX, newY, oldZ);
```

fmod works like modulo on integers and returns the remainder of the division. If we use fmod with the tile size, so every tile on the checkerboard. We basically move the checkerboard along with the character and just displace it a little bit within range [0, tileSize]. And we just do that every frame.

**Review by amzn-jillich — APPROVED** [review]

File: Gems/AtomLyIntegration/EMotionFXAtom/Code/Tools/EMStudio/AnimViewportRenderer.h [hunk 2]
```diff
@@ -91,6 +94,8 @@ namespace EMStudio
         const RenderOptions* m_renderOptions;
 
         const float DefaultFrustumDimension = 128.0f;
+        const float BoundMaxDistance = 150.0f;
```
Comment (amzn-jillich) [inline Gems/AtomLyIntegration/EMotionFXAtom/Code/Tools/EMStudio/AnimViewportRenderer.h:97]: That's 150 from 500 tiles, don't we see the ground plane popping in the horizon with such a high number? What about doing something like 10 or so? Personally, I'd remove the condition entirely and just re-position it every frame.

File: Gems/AtomLyIntegration/EMotionFXAtom/Code/Tools/EMStudio/AnimViewportRenderer.cpp [hunk 3]
```diff
@@ -224,6 +224,22 @@ namespace EMStudio
...
+        AZ::TransformBus::EventResult(groundPos, m_groundEntity->GetId(), &AZ::TransformBus::Events::GetWorldTranslation);
+
+        const AZ::Vector3 characterPos = GetCharacterCenter();
+        if (AZStd::abs(characterPos.GetX() - groundPos.GetX()) > BoundMaxDistance ||
+            AZStd::abs(characterPos.GetY() - groundPos.GetY()) > BoundMaxDistance)
+        {
+            const float tileOffsetX = AZStd::fmod(characterPos.GetX(), TileSize);
+            const float tileOffsetY = AZStd::fmod(characterPos.GetX(), TileSize);
+            const AZ::Vector3 newGroundPos(characterPos.GetX() - tileOffsetX, characterPos.GetY() - tileOffsetY, groundPos.GetZ());
+            AZ::TransformBus::Event(m_groundEntity->GetId(), &AZ::TransformBus::Events::SetWorldTranslation, newGroundPos);
```
Comment (amzn-jillich) [inline Gems/AtomLyIntegration/EMotionFXAtom/Code/Tools/EMStudio/AnimViewportRenderer.cpp:239]: Awesome!

File: Gems/AtomLyIntegration/EMotionFXAtom/Code/Tools/EMStudio/AnimViewportRenderer.h [hunk 4]
```diff
@@ -91,6 +94,8 @@ namespace EMStudio
...
+        const float TileSize = 1.0f;
```
Comment (amzn-jillich) [inline Gems/AtomLyIntegration/EMotionFXAtom/Code/Tools/EMStudio/AnimViewportRenderer.h:98]: Nit: You might want to make them `static constexpr` and give them a `s_` prefix.

**Review by amzn-jillich — APPROVED** [review]

**Comment by amzn-jillich** [comment]
Resolves #7646

**Review by lsemp3d — APPROVED** [review]

**Pull request merged**